From 7afaa0057963dcf584d3e9ef07aa09f4be89c126 Mon Sep 17 00:00:00 2001
From: Amilendra Kodithuwakku <amilendra.kodithuwakku@arm.com>
Date: Thu, 29 Jan 2026 16:45:01 +0000
Subject: Initialize SME during C-runtime initialization (crt0)

Check if SME or SME2 is enabled and do the following

1. Clear TPIDR2_EL0 [1]
The value of TPIDR2_EL0 resets to an architecturally UNKNOWN value
at warm reset.
This means we need to clear it manually at startup.

2. Initialise SME Streaming SVE Vector length to 2048 bits
Initialize the SME control register SMCR_EL3 [2] to set the
effective streaming SVE vector length to the maximum supported
value of 2048 bits.

[1] https://developer.arm.com/documentation/ddi0601/2025-09/AArch64-Registers/TPIDR2-EL0--EL0-Read-Write-Software-Thread-ID-Register-2?lang=en
[2] https://developer.arm.com/documentation/ddi0601/2025-09/AArch64-Registers/SMCR-EL3--SME-Control-Register--EL3-?lang=en
---
 picocrt/machine/aarch64/crt0.S | 23 +++++++++++++++++++++++
 1 file changed, 23 insertions(+)

diff --git a/picocrt/machine/aarch64/crt0.S b/picocrt/machine/aarch64/crt0.S
index 8b57ecf697..351c4877c4 100644
--- a/picocrt/machine/aarch64/crt0.S
+++ b/picocrt/machine/aarch64/crt0.S
@@ -41,6 +41,9 @@
 
 #define MMU_INVALID_FLAGS       0
 
+#define FEAT_SME_1      (1LL << 24)
+#define FEAT_SME_2      (1LL << 25)
+
   .macro start_page_table
   .section .rodata
   .global __identity_page_table
@@ -149,6 +152,26 @@ _start:
 #endif // __ARM_FP
 #endif
 
+	/* Determine if SME or SME2 is available */
+	mrs     x0, id_aa64pfr1_el1
+	tst     x0, (FEAT_SME_1|FEAT_SME_2)
+	b.eq    .Lnosme
+
+	/* Clear TPIDR2_EL0 */
+	msr     S3_3_C13_C0_5, xzr
+	isb
+
+	/* set up Streaming SVE Vector Length (SVL) in SMCR_EL3 (4 LSB). */
+	mov     x2, #0xF
+
+	/* Try to set the maximum value supported by the architecture (2048). */
+	mrs     x1, s3_6_c1_c2_6 /* mrs  x1, smcr_el3 */
+	bfi     x1, x2, 0, 4
+	msr     s3_6_c1_c2_6, x1 /* msr  smcr_el3, x1 */
+	isb
+
+.Lnosme:
+
 	/* Jump into C code */
 	bl _cstart
   .size _start, .-_start
-- 
2.43.0

